src/components/ChannelPreview/hooks/useMessageDeliveryStatus.ts

```diff
@@ -0,0 +1,85 @@
+import { useCallback, useEffect, useState } from 'react';
+
+import type { Channel, Event } from 'stream-chat';
+
+import type { DefaultStreamChatGenerics } from '../../../types/types';
+import type { StreamMessage } from '../../../context';
+import { useChatContext } from '../../../context';
+
+export enum MessageDeliveryStatus {
+  DELIVERED = 'delivered',
+  READ = 'read',
+}
+
+type UseMessageStatusParamsChannelPreviewProps<
+  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
+> = {
+  channel: Channel<StreamChatGenerics>;
+  /** The last message received in a channel */
+  lastMessage?: StreamMessage<StreamChatGenerics>;
+};
+
+export const useMessageDeliveryStatus = <
+  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
+>({
+  channel,
+  lastMessage,
+}: UseMessageStatusParamsChannelPreviewProps<StreamChatGenerics>) => {
+  const { client } = useChatContext();
+  const [messageDeliveryStatus, setMessageDeliveryStatus] = useState<MessageDeliveryStatus>();
+
+  const isOwnMessage = useCallback(
+    (message?: StreamMessage<StreamChatGenerics>) =>
+      client.user && message?.user?.id === client.user.id,
+    [client],
+  );
+
+  useEffect(() => {
+    const lastMessageIsOwn = isOwnMessage(lastMessage);
+    if (!lastMessage?.created_at || !lastMessageIsOwn) return;
+
+    const lastMessageCreatedAtDate =
+      typeof lastMessage.created_at === 'string'
+        ? new Date(lastMessage.created_at)
+        : lastMessage.created_at;
+
+    const channelReadByOthersAfterLastMessageUpdate = Object.values(channel.state.read).some(
+      ({ last_read: channelLastMarkedReadDate, user }) => {
+        const ignoreOwnReadStatus = client.user && user.id !== client.user.id;
+        return ignoreOwnReadStatus && lastMessageCreatedAtDate < channelLastMarkedReadDate;
+      },
+    );
+
+    setMessageDeliveryStatus(
+      channelReadByOthersAfterLastMessageUpdate
+        ? MessageDeliveryStatus.READ
+        : MessageDeliveryStatus.DELIVERED,
+    );
+  }, [channel.state.read, client, isOwnMessage, lastMessage?.created_at]);
```

Comment:
We don't have `lastMessage` in the list of dependencies here. Is that intentional?

Reply:
Theoretically, we need the whole message, as it is used in `isOwnMessage` that internally checks for `message.user.id`.
Also, we should check whether there are accidental mutations of `lastMessage` that may break the `useEffect`.

Reply:
True

Reply:
Fixed